crates/lib/src/compiler/libquil.rs

```diff
@@ -0,0 +1,349 @@
+    /// Error when casting u64 to u32
+    #[error("error when casting u64 to u32: {0}")]
+    U64Truncation(#[from] TryFromIntError),
+    /// Error when creating a CString
+    #[error("error when creating CString: {0}")]
+    CString(#[from] NulError),
+}
+
+impl From<Error> for quilc::Error {
+    fn from(error: Error) -> Self {
+        quilc::Error::QuilcCompilation(error.to_string())
+    }
+}
+
+impl From<libquil_sys::quilc::CompilationMetadata> for NativeQuilMetadata {
+    fn from(value: libquil_sys::quilc::CompilationMetadata) -> Self {
+        NativeQuilMetadata {
+            final_rewiring: value
+                .final_rewiring
+                .iter()
+                .map(|r| u64::from(*r))
+                .collect::<Vec<_>>(),
+            gate_depth: value.gate_depth.map(u64::from),
+            gate_volume: value.gate_volume.map(u64::from),
+            multiqubit_gate_depth: value.multiqubit_gate_depth.map(u64::from),
+            program_duration: value.program_duration,
+            program_fidelity: value.program_fidelity,
+            topological_swaps: value.topological_swaps.map(u64::from),
+            qpu_runtime_estimation: value.qpu_runtime_estimation,
+        }
+    }
+}
+
+/// A libquil client providing Quilc functionality
+#[derive(Debug, Clone, Copy)]
+pub struct Client;
+
+impl quilc::Client for Client {
+    fn compile_program(
+        &self,
+        quil: &str,
+        isa: quilc::TargetDevice,
+        options: quilc::CompilerOpts,
+    ) -> Result<quilc::CompilationResult, quilc::Error> {
+        let program = libquil_sys::quilc::Program::from_str(quil).map_err(Error::from)?;
+        let isa = serde_json::to_string(&isa).map_err(Error::from)?;
+        let chip = libquil_sys::quilc::Chip::from_str(&isa).map_err(Error::from)?;
+
+        let compilation_result = if options.protoquil.unwrap_or_default() {
+            libquil_sys::quilc::compile_protoquil(&program, &chip)
+        } else {
+            libquil_sys::quilc::compile_program(&program, &chip)
+        }
+        .map_err(Error::from)?;
+
+        let program = compilation_result
+            .program
+            .to_string()
+            .map_err(Error::from)?
+            .parse()
+            .map_err(Error::from)?;
+        Ok(quilc::CompilationResult {
+            program,
+            native_quil_metadata: compilation_result.metadata.map(Into::into),
+        })
+    }
+
+    fn get_version_info(&self) -> Result<String, quilc::Error> {
+        Ok(libquil_sys::quilc::get_version_info()
+            .map_err(Error::from)?
+            .version)
+    }
+
+    fn conjugate_pauli_by_clifford(
+        &self,
+        request: quilc::ConjugateByCliffordRequest,
+    ) -> Result<quilc::ConjugatePauliByCliffordResponse, quilc::Error> {
+        let pauli_terms = request
+            .pauli
+            .symbols
+            .into_iter()
+            .map(CString::new)
+            .collect::<Result<Vec<_>, _>>()
```

Comment:
I'd only recommend using `itertools` if it is already present as a dependency -- it's not worth adding just for this

Reply:
Also the above example @BatmanAoD gave actually won't work because of the `?` -- setting the type outside of the turbofish is stating what is expected _after_ the `?`. This should work, though:

```rust
let pauli_terms: Vec<_> = request
    ...
    .collect::<Result<_, _>>()?;
```

Reply:
I don't really mind the turbofish syntax. Kinda like it, and not only because it's called "turbofish" 🤷🏽

Reply:
That said, the `Vec<_>` _is_ unnecessary in the referenced code.